File: 0078-subsets/0078-subsets.py
class Solution:
    def helper(self, ind, nums, temp, ans):
        if(ind == len(nums)):
            ans.append(temp[:])
            return

        if(ind > len(nums)):
            return
        
        # take
        temp.append(nums[ind])
        self.helper(ind+1, nums, temp, ans)
        temp.pop()

        # notTake
        self.helper(ind+1, nums, temp, ans)
    
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans = []

        self.helper(0, nums, [], ans)

        return ans;

File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    void helper(int ind, vector<vector<int>> &ans, vector<int> ds, int n, vector<int> nums){
        if(ind==n){
            ans.push_back(ds);
            return;
        }
        // NOT TAKE
        helper(ind+1, ans, ds, n, nums);

        // TAKE
        ds.push_back(nums[ind]);
        helper(ind+1, ans, ds, n, nums);
        // ds.pop_back();
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> ds;
        int n = nums.size();

        helper(0, ans, ds, n, nums);
        return ans;
    }
};